### ISSUE TITLE
Validation fails with SSLException after CA certificate is "rolled-over"

### ISSUE THREAD

**Issue by paulmillar** [eu-emi/canl-java#107]
CA certificate "roll-over" is when a CA certificate (e.g,. from an intermediate CA) is updated with a replacement CA.

One way to do this is to use the same public-private key-pair as the credential from which to generate the updated CA certificate.  This should allow already issued EECs to validate against the intermediate CA: the hash of the EEC was signed by the previous CA credential's private key, which is the same as the updated CA credential.

The most recent roll-over happened with the release of IGTF trust-anchor distribution v1.109.  With the upgrade to v1.109, the intermediate CA `QuoVadis-Grid-ICA-G2` was updated with a replacement.  Here are the details of the certificate in v1.108 distribution:

```
paul@sprocket:~$ openssl x509 -in QuoVadis-Grid-ICA-G2.pem -noout -text
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            10:b6:2f:34:5a:61:eb:61:84:a7:b1:06:c8:05:7e:94:53:e2:07:d5
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: C = BM, O = QuoVadis Limited, CN = QuoVadis Root CA 2
        Validity
            Not Before: May 23 15:27:02 2016 GMT
            Not After : May 23 15:27:02 2026 GMT
        Subject: C = BM, O = QuoVadis Limited, CN = QuoVadis Grid ICA G2
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                RSA Public-Key: (4096 bit)
                Modulus:
                    00:d8:37:0f:c3:68:c1:ee:c9:d5:a9:1a:b7:ff:e7:
[...]
                    89:0f:69
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Basic Constraints: critical
                CA:TRUE, pathlen:0
            X509v3 Certificate Policies: 
                Policy: X509v3 Any Policy

            Authority Information Access: 
                OCSP - URI:http://ocsp.quovadisglobal.com
                CA Issuers - URI:http://trust.quovadisglobal.com/qvrca2.crt

            X509v3 Key Usage: critical
                Certificate Sign, CRL Sign
            X509v3 Authority Key Identifier: 
                keyid:1A:84:62:BC:48:4C:33:25:04:D4:EE:D0:F6:03:C4:19:46:D1:94:6B

            X509v3 CRL Distribution Points: 

                Full Name:
                  URI:http://crl.quovadisglobal.com/qvrca2.crl

            X509v3 Subject Key Identifier: 
                B7:5D:75:82:97:2C:B2:FB:6E:FE:1C:19:50:E0:2B:6F:E4:DF:BA:30
    Signature Algorithm: sha256WithRSAEncryption
         12:71:53:22:09:31:a1:ec:2b:d4:41:c2:d5:4b:5e:d9:4c:ae:
[...]
         f5:58:e9:67:91:56:2b:92
paul@sprocket:~$ 
```

Here is the information about the updated certificate:

```
paul@sprocket:~$ openssl x509 -in /etc/grid-security/certificates/QuoVadis-Grid-ICA-G2.pem -noout -text
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            2b:af:c5:f0:a8:d2:13:40:9d:a4:2e:76:1f:61:9a:c9:19:84:bc:d4
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: C = BM, O = QuoVadis Limited, CN = QuoVadis Root CA 2
        Validity
            Not Before: Sep 22 19:21:12 2020 GMT
            Not After : May 23 15:27:02 2026 GMT
        Subject: C = BM, O = QuoVadis Limited, CN = QuoVadis Grid ICA G2
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                RSA Public-Key: (4096 bit)
                Modulus:
                    00:d8:37:0f:c3:68:c1:ee:c9:d5:a9:1a:b7:ff:e7:
[...]
                    89:0f:69
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Basic Constraints: critical
                CA:TRUE, pathlen:0
            X509v3 Authority Key Identifier: 
                keyid:1A:84:62:BC:48:4C:33:25:04:D4:EE:D0:F6:03:C4:19:46:D1:94:6B

            Authority Information Access: 
                CA Issuers - URI:http://trust.quovadisglobal.com/qvrca2.crt
                OCSP - URI:http://ocsp.quovadisglobal.com

            X509v3 Certificate Policies: 
                Policy: X509v3 Any Policy
                  CPS: https://www.quovadisglobal.com/repository

            X509v3 Extended Key Usage: 
                TLS Web Client Authentication, TLS Web Server Authentication
            X509v3 CRL Distribution Points: 

                Full Name:
                  URI:http://crl.quovadisglobal.com/qvrca2.crl

            X509v3 Subject Key Identifier: 
                B7:5D:75:82:97:2C:B2:FB:6E:FE:1C:19:50:E0:2B:6F:E4:DF:BA:30
            X509v3 Key Usage: critical
                Certificate Sign, CRL Sign
    Signature Algorithm: sha256WithRSAEncryption
         10:11:19:91:93:ea:db:79:a6:76:65:dc:8d:e2:00:73:35:3b:
[...]
         e3:b4:82:ab:07:69:90:6e
paul@sprocket:~$ 
```

Note that the old and new certificates have the same public key (the RSA modulus and exponent).

Many of the extensions are the same (albeit some with a different order).  The only significant differences are the "Not Before" date has been updated, the new certificate Policies now include a `CPS`, and there is now an Extended Key Usage extension.

After an application using CANL experiences this CA roll-over, certificates issued by this intermediate CA start to fail.  The error is somewhat striking: the validation fails with SSLException, but without any message to describe what went wrong.

I can confirm that restarting the application "fixes" this issue.  The fresh application "sees" only the updated CA certificate and EECs issued by the CA validate as expected.

The problem is remarkably similar to the one described in #94. This leaves me wondering whether the fix for that issue was only partially effective.

**Comment by paulmillar** [comment]
I can add some more information.

The fix for #94 (commit 7765bbd) was committed to CaNL master branch.  However, it was not back-ported to the supported branches (i.e., 1.5).  Therefore, at that time, there was no released version of CaNL that contained the bug-fix.

The master branch was eventually released as v1.6.0 (a year later) but unfortunately I wasn't aware of this release, so didn't patch dCache to use the newer version.

Therefore, it's not that surprising that the problem wasn't fixed and the same problem reoccurred.

**Comment by rptaylor** [comment]
Which dcache service needs to be restarted, just gplazma?

**Comment by maarten-litmaath** [comment]
Hi Ryan,
in a separate thread on this matter, Paul wrote:

In the meantime, the only solution (that I know of) is to restart
dCache.  Unfortunately, this requires all door nodes, all pool nodes and
nodes running gPlazma to be restarted.  For most sites, this is
basically restarting all of dCache.

**Comment by paulmillar** [comment]
Since this ticket has diverged slightly into a discussion on dCache support, perhaps I should clarify.

dCache uses CaNL for _all_ X.509 certificate verification: if a person, an agent  (operating on behalf of a person; e.g., FTS), or a host presents an X.509 certificate chain to dCache then dCache uses CaNL to validate that chain.

The problem reported so far have been about HTTP-TPC, where dCache connects to the remote host as an HTTP client.  The remote host presents an X.509 credential (the host cert.) which dCache validates using CaNL (see above statement).  As a result of the subject of this CaNL issue, this validation fails because the certificate was issued by QuoVadis, which recently "rolled over" their intermediate CA certificate.

This specific problem is fixed by restarting all pool nodes (and only the pool nodes) involved with HTTP-TPC.  This is because it is only a pool node that connects to the remote server and, therefore, validates the QuoVadis-issued host certificate.  Any pool nodes that are never used for HTTP-TPC may be left alone.

However, QuoVadis may also issue personal certificates.  I don't have a specific example of this, but I don't see any reason why they would not issue personal certificate so I'm assuming they do.  Such personal certificates would also fail, for the same reason as the host certificates fail.  In this case, the validation step happens twice: once in the door (during the TLS handshake) and once in gPlazma (mapping the user's identity).  Therefore, the doors and gPlazma must also be restarted to avoid problems when users with  QuoVadis-issued certificate use dCache.

Only doors that accept X.509-based authentication need to be restarted; however, for most sites, that's basically all of them.

The TL;DR is: restart all nodes in dCache.

**Comment by golbi** [comment]
@paulmillar  while limited on time I'm trying to make some small refresh of canl library and knock down the most important problems. This one looks like fixed, but not in some old 1.x release chain. Can you please write whether this issue is relevant anymore? And if yes provide pull request to backport the aforementioned change? I can try to release the old version, but anyway upgrading would be much appreciated.

**Comment by paulmillar** [comment]
The short answer is that (I believe) this is no longer a problem for dCache.  The issue may be closed.

As an aside, it looks like [my comment above](https://github.com/eu-emi/canl-java/issues/107#issuecomment-769867914) uses the wrong CaNL version numbers.  Where it says "1.5" should have been 2.5 (referring to CaNL v2.5.0/v2.5.1) and "1.6" should have been 2.6 (as in CaNL v2.6.0).

CaNL v2.6.0 contains [commit 7765bbdc](https://github.com/eu-emi/canl-java/commit/7765bbdcf2a5d736c739e9032bf977885d18d72e), which fixes this issue.

We've updated dCache so it is now using CaNL v2.6.0 for all supported branches, except for dCache v6.2 (for which we're dropping support this summer).  Therefore, this problem is now fixed, provided people are running an up-to-date version of dCache.
